clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -1184,15 +1186,19 @@ private TxnOffsetCommitHandler txnOffsetCommitHandler(TransactionalRequestResult
     }
 
     private TransactionalRequestResult handleCachedTransactionRequestResult(
-            Supplier<TransactionalRequestResult> transactionalRequestResultSupplier,
-            State targetState) {
+        Supplier<TransactionalRequestResult> transactionalRequestResultSupplier,
+        State transientState
+    ) {
         ensureTransactional();
 
-        if (pendingResult != null && currentState == targetState) {
+        if (pendingResult != null && !pendingResult.isAcked()) {
             TransactionalRequestResult result = pendingResult;
-            if (result.isCompleted())
+            if (result.isCompleted() && priorState == transientState) {
                 pendingResult = null;
-            return result;
+                return result;
+            } else if (currentState == transientState) {
+                return result;
```

Comment:
Why don't we set `pendingResult` to `null` when `isCompleted()` is `true` here? I suppose that we assume that the `currentState` which transition to the next state whenever the response is received so we should never have a completed request here. Therefore, the "prior state" check will handle all the cases. Is my reasoning correct?

Reply:
The logic was pretty confusing, so I attempted to simplify it in the latest commit. The need to ensure that the result gets communicated to the user definitely makes this code more awkward.